### PR TITLE
[Bug](datev2) Fix bug when cast datev2 to date

### PR DESCRIPTION
**Pull request by Gabriel39**
# Proposed changes

Bug:
image

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/function_cast.h [hunk 1]
```diff
@@ -194,17 +194,16 @@ struct ConvertImpl {
                     } else if constexpr (IsTimeType<ToDataType>) {
                         if constexpr (IsDateTimeType<ToDataType> && IsDateV2Type<FromDataType>) {
                             DataTypeDateV2::cast_to_date_time(vec_from[i], vec_to[i]);
-                        } else if constexpr (IsDateTimeV2Type<ToDataType> &&
-                                             IsDateV2Type<FromDataType>) {
+                        } else if constexpr (IsDateType<ToDataType> && IsDateV2Type<FromDataType>) {
```
Comment (carlvinhust2012) [inline be/src/vec/functions/function_cast.h:197]: so here is check the cast from 'DateV2Type' to 'DateType'?

Comment (Gabriel39) [reply be/src/vec/functions/function_cast.h:197]: yep

**Review by HappenLee — APPROVED** [review]
LGTM

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.69 seconds
 load time: 515 seconds
 storage size: 17170835939 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230203102626_clickbench_pr_90076.html

File: regression-test/suites/datatype_p0/date/test_date_exprs.groovy [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    qt_sql1 """ select dt
+             from
+             (
+             select cast(k1 as datev2) as dt
+             from ${tbName}
+             ) r; """
+    qt_sql2 """ select dt
+             from
+             (
+             select cast(k1 as datev2) as dt
```
Comment (morningman) [inline regression-test/suites/datatype_p0/date/test_date_exprs.groovy:36]: select cast(k1 as date) as dt

**Pull request merged**